### PR TITLE
fix: add missing tests, correct calls to DictWrapper constructor and improve metrics type hints

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:**

## Description of changes:

Changes:
* validator - Add test for the validator decator with an envelope
* data_classes - Call DictWrapper constructor with the nested dict
* metrics - Some minor type hint improvements

**Checklist**

 

* [X] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [X] Update tests
* [X] Update docs
* [X] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/metrics/base.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def serialize_metric_set(self, metrics: Dict = None, dimensions: Dict = None, me
         logger.debug({"details": "Serializing metrics", "metrics": metrics, "dimensions": dimensions})
 
         metric_names_and_units: List[Dict[str, str]] = []  # [ { "Name": "metric_name", "Unit": "Count" } ]
-        metric_names_and_values: Dict[str, str] = {}  # { "metric_name": 1.0 }
+        metric_names_and_values: Dict[str, int] = {}  # { "metric_name": 1.0 }
```
Comment (heitorlessa) [inline aws_lambda_powertools/metrics/base.py:188]:
```suggestion
        metric_names_and_values: Dict[str, float] = {}  # { "metric_name": 1.0 }
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Thanks for those changes @michaelbrewer, and for catching an issue with the data_classes utilities.

~~Could you update description to reflect what changed here? Title reflects validator test only.~~

One minor issue with type hint as it should've been float over int due to value, and numeric tower also catches int as part of float

**Review by michaelbrewer — COMMENTED** [review]
Corrected

File: aws_lambda_powertools/metrics/base.py [hunk 2]
```diff
@@ -125,7 +125,7 @@ def add_metric(self, name: str, unit: Union[MetricUnit, str], value: float):
             raise MetricValueError(f"{value} is not a valid number")
 
         unit = self.__extract_metric_unit_value(unit=unit)
-        metric = self.metric_set.get(name, defaultdict(list))
+        metric: dict = self.metric_set.get(name, defaultdict(list))
```
Comment (heitorlessa) [inline aws_lambda_powertools/metrics/base.py:128]:
```suggestion
        metric: Dict = self.metric_set.get(name, defaultdict(list))
```

nitpick

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**